backend/pkg/api/data_access/notifications.go

```diff
@@ -77,11 +100,534 @@ func (d *DataAccessService) UpdateNotificationSettingsClients(ctx context.Contex
+	settings := slices.Collect(maps.Values(settingsMap))
+
+	// -------------------------------------
+	// Sort
+	// Each row is uniquely defined by the dashboardId, groupId, and isAccountDashboard so the sort order is DashboardName/GroupName => DashboardId => GroupId => IsAccountDashboard
+	var primarySortParam func(resultEntry *NotificationSettingsDashboardsInfo) string
+	switch colSort.Column {
+	case enums.NotificationSettingsDashboardColumns.DashboardName:
+		primarySortParam = func(resultEntry *NotificationSettingsDashboardsInfo) string { return resultEntry.DashboardName }
+	case enums.NotificationSettingsDashboardColumns.GroupName:
+		primarySortParam = func(resultEntry *NotificationSettingsDashboardsInfo) string { return resultEntry.GroupName }
+	default:
+		return nil, nil, fmt.Errorf("invalid sort column for notification subscriptions: %v", colSort.Column)
+	}
+	sort.Slice(settings, func(i, j int) bool {
+		if isReverseDirection {
+			if primarySortParam(settings[i]) == primarySortParam(settings[j]) {
+				if settings[i].DashboardId == settings[j].DashboardId {
+					if settings[i].GroupId == settings[j].GroupId {
+						return settings[i].IsAccountDashboard
+					}
+					return settings[i].GroupId > settings[j].GroupId
+				}
+				return settings[i].DashboardId > settings[j].DashboardId
+			}
+			return primarySortParam(settings[i]) > primarySortParam(settings[j])
+		} else {
+			if primarySortParam(settings[i]) == primarySortParam(settings[j]) {
+				if settings[i].DashboardId == settings[j].DashboardId {
+					if settings[i].GroupId == settings[j].GroupId {
+						return settings[j].IsAccountDashboard
+					}
+					return settings[i].GroupId < settings[j].GroupId
+				}
+				return settings[i].DashboardId < settings[j].DashboardId
+			}
+			return primarySortParam(settings[i]) < primarySortParam(settings[j])
+		}
+	})
+
+	// -------------------------------------
+	// Convert to the final result format
+	for _, setting := range settings {
+		result = append(result, t.NotificationSettingsDashboardsTableRow{
+			IsAccountDashboard: setting.IsAccountDashboard,
+			DashboardId:        setting.DashboardId,
+			GroupId:            setting.GroupId,
+			GroupName:          setting.GroupName,
+			Settings:           setting.Settings,
+			ChainIds:           setting.ChainIds,
+		})
+	}
+
+	// -------------------------------------
+	// Paging
+
+	// Find the index for the cursor and limit the data
+	if currentCursor.IsValid() {
+		for idx, row := range settings {
+			if row.DashboardId == currentCursor.DashboardId &&
+				row.GroupId == currentCursor.GroupId &&
+				row.IsAccountDashboard == currentCursor.IsAccountDashboard {
+				result = result[idx+1:]
+				break
+			}
+		}
+	}
+
+	// Flag if above limit
+	moreDataFlag := len(result) > int(limit)
+	if !moreDataFlag && !currentCursor.IsValid() {
+		// No paging required
+		return result, &paging, nil
+	}
+
+	// Remove the last entries from data
+	if moreDataFlag {
+		result = result[:limit]
+	}
```

Comment:
Yes, I always try to get the LIMIT and ORDER BY into the query when possible.
Currently the `users_val_dashboards` could have these since we show all dashboards but it would no longer work if we have the account dashboards query in a separate routine.
Perhaps once these should be implemented and their db structure is defined we can unite them efficiently into one query and add LIMIT/ORDER BY.

However if you want the `users_val_dashboards` query adjusted now I could also do it.

Reply:
Sounds like a good approach, don't think fully optimizing this particular one is too important for now. So agreed, might do it when we add the account dashboards